app/models/host.rb

```diff
@@ -59,6 +59,17 @@ def readonly?
     true
   end
 
+  def self.os_version_query(path, values, equal = true)
+    query = equal ? :in : :not_in
+    raise ArgumentError unless %i[major minor].include?(path)
+
+    Arel::Nodes::NamedFunction.new('CAST', [Arel::Nodes::InfixOperation.new(
+      '->',
+      arel_table[:system_profile],
+      Arel::Nodes::SqlLiteral.new("'operating_system'->'#{path}'")
+    ).as('int')]).send(query, values)
```

Comment:
What would happen with hosts that have empty string on major or minor OS version.

Reply:
I don't really understand the question :thinking:

Reply:
The db. This queries JSON in `system_profile` field, populated by Inventory via Cyndi. Some hosts either might lack `operating_system.minor` (or major) values and can have it as an empty string. In such cases this would fall app art like house of cards. :grin:

Reply:
The `CAST` in pg doesn't convert `NULL` to `0` so I don't see how it would happen:

```
SELECT CAST(NULL as int), CAST(0 as int);
 int4 | int4 
------+------
      |    0
(1 row)
```

Reply:
```
# SELECT CAST('' as int);
ERROR:  invalid input syntax for type integer: ""
LINE 1: SELECT CAST('' as int);
```

Reply:
I resolved it with an additional layer of casting+nullif